source/common/quic/envoy_quic_utils.h

```diff
@@ -45,16 +45,26 @@ class HeaderValidator {
   validateHeader(const std::string& header_name, absl::string_view header_value) = 0;
 };
 
+static constexpr absl::string_view invalid_underscore = "http3.unexpected_underscore";
+static constexpr absl::string_view too_many_http_headers = "http3.too_many_http_headers";
+
 // The returned header map has all keys in lower case.
 template <class T>
-std::unique_ptr<T> quicHeadersToEnvoyHeaders(const quic::QuicHeaderList& header_list,
-                                             HeaderValidator& validator) {
+std::unique_ptr<T>
+quicHeadersToEnvoyHeaders(const quic::QuicHeaderList& header_list, HeaderValidator& validator,
+                          uint32_t max_headers_allowed, absl::string_view& details) {
```

Comment:
Is it because QuicHeaderList doesn't have expose size() interface that we need max_headers_allowed to be passed into this function?

I slightly prefer checking it in QUICHE and have something like OnHeadersTooLarge() or simply add QuicHeaderList::size() instead of iterating over the header list again.

Reply:
SGTM.